### ISSUE TITLE
Allow to limit the "in-memory" chunks for Ingester

### ISSUE THREAD

**Issue by TeddyAndrieux** [grafana/loki#5721]
**Is your feature request related to a problem? Please describe.**
If I configure Loki to store on the filesystem and the disk used get full, since it cannot write on the disk Loki will keep receiving data and fill the memory until it gets OOM killed (or the machine just crash).

**Describe the solution you'd like**
To me, we should be able to configure the Loki ingester "in memory max size" (the maximum size the "in-memory" chunk should take) so that once Loki reaches this "in-memory max size" he just reject new data with the right return code "retry-later". 

**Describe alternatives you've considered**
If Loki is deployed as a Pod in Kubernetes we could just set a memory limit, but it's not a good solution IMHO since it means once Loki reach this limit, the Pod just get OOM killed and we likely lose all "in-memory" data

**Additional context**

**Comment by TeddyAndrieux** [comment]
Also maybe Loki should expose 2 different endpoints:
- `/health` that basically do what the `/ready` endpoint does today
- `/ready` that also check that we can actually write chunks (or maybe that we can just write in the WAL)

**Comment by chaudum** [comment]
> If I configure Loki to store on the filesystem and the disk used get full, since it cannot write on the disk Loki will keep receiving data and fill the memory until it gets OOM killed (or the machine just crash).

To me, these are two independent areas:
1. Disk filling up and WAL cannot be written any more. This usually happens first, before chunks are filling up memory.
2. Chunks filling up memory and causing process to run OOM.

Both should be handled independently.

> To me, we should be able to configure the Loki ingester "in memory max size" (the maximum size the "in-memory" chunk should take) so that once Loki reaches this "in-memory max size" he just reject new data with the right return code "retry-later".

Instead of manually configuring the limit, what about having a "dynamic" limit based on the available memory?

**Comment by dannykopping** [comment]
> Also maybe Loki should expose 2 different endpoints:
> 
> * `/health` that basically do what the `/ready` endpoint does today
> * `/ready` that also check that we can actually write chunks (or maybe that we can just write in the WAL)

Agreed, the health/ready check should take the WAL appendability into account

**Comment by TeddyAndrieux** [comment]
> To me, these are two independent areas:
> 
> 1. Disk filling up and WAL cannot be written any more. This usually happens first, before chunks are filling up memory.
> 2. Chunks filling up memory and causing process to run OOM.
> 
> Both should be handled independently.

I totally agree 1. is not the point for this feature request (Ideally could be cool to have a retention size so that old chunk get automatically cleaned when we reach this size) but the point for this feature request is really about 2. 

> Instead of manually configuring the limit, what about having a "dynamic" limit based on the available memory?

I'm not entirely sure what you mean by "dynamic", you mean like we set a percentage so that Ingester checks all memory and "compute" his limit ? Fine for me as well

>  if we define a threshold of disk fullness at which we start to reject writes, that'll save the ingester from having a large increase in memory usage anyway.

Indeed, like a retention size

**Comment by dannykopping** [comment]
I don't think we should change the way we buffer chunks in memory, but rather address the underlying cause - which is the WAL append failure behaviour.

**Comment by TeddyAndrieux** [comment]
From what I see it's not the same root cause, if you are able to write the WAL but not able to write data at the end you will still consume all the memory for chunks. 
I think in most cases when you use filesystem storage you likely use the same disk for WAL and data so it's related but it may not, in some context.

I don't know if the same may happen with some other config like using Cassandra/s3 or whatever else storage, like Loki unable to write to his storage so it keep all chunks in memory (even if it can write WAL)

**Comment by dannykopping** [comment]
I was wrong about the behaviour of the ingester, sorry for the confusion.

The ingester will receive samples into memory and then produce "chunks", which are then flushed to object storage. When this happens (and after `chunk_idle_period`), the chunk is purged from memory. I believe the only way an ingester will fill up its memory is if the chunks cannot be flushed successfully to the chunk store fast enough. More details [here](https://taisho6339.gitbook.io/grafana-loki-deep-dive/ingestion-process/buffer-log-chunks).

Here's what I think we could do moving forward:

1. Have a background process which checks the "health" of the WAL disk - i.e. is it full (or nearly full)? Before appending the samples into the head chunk, check if the WAL is ready. If not, reject writes until the WAL is ready again
  - This is a bit complex because to determine the mountpath on which the WAL resides is a bit tricky and not portable across OSes
2. If the chunks cannot flush to the chunk store, measure the size of the chunk flush queue and if it becomes dangerously long, start rejecting writes

> 2- A way to set a memory limit (once Loki reach this limit he just answer 5xx to every new POST (and stop actually reading the WAL))

I'm not yet convinced why this would be an idea worth exploring, but I'm open to suggestions. Solving the problem that grows ingester memory in an unacceptable way (chunks cannot flush to storage) seems like the preferable course of action first.

**Comment by dannykopping** [comment]
Also adding this here as it's related: https://github.com/grafana/loki/pull/3136/files#

**Comment by frittentheke** [comment]
I just noticed the `/run/loki` folder filling most of the memory due to S3 being unreachable.
An enforced limit potentially related to the available space should absolutely be added to Loki to not push the machine into OOM.
Also just returning a "503 Service Unavailable" until the issue is resolved should be enough to have the clients buffer and try again.

**Comment by Nachtfalkeaw** [comment]
I think the issue is still existent in 3.7.3

If s3 is unreachable then memory fills Up. There should be a memory Limit which makes the ingester send 429 to the Clients in ingest path to make Sure it does Not Run Out of memory and Clients Buffer.

In the other Hand read path should be still available.
